packages/bvaughn-architecture-demo/src/suspense/PointsCache.ts

```diff
@@ -0,0 +1,66 @@
+import { ExecutionPoint, SessionId } from "@replayio/protocol";
+import { unstable_getCacheForType as getCacheForType } from "react";
+import { ReplayClientInterface } from "../ReplayClient";
+
+import { Record, STATUS_PENDING, STATUS_REJECTED, STATUS_RESOLVED, Wakeable } from "../types";
+import { createWakeable } from "../utils/suspense";
+
+// TODO We could add some way for external code (in the client adapter) to pre-populate this cache with known points.
+// For example, when we get Paints they have a corresponding point (and time) which could be pre-loaded here.
+
+type TimeToRecordMap = Map<number, Record<ExecutionPoint>>;
+
+function createMap(): TimeToRecordMap {
+  return new Map();
+}
+
+function getRecordMap(): TimeToRecordMap {
+  return getCacheForType(createMap);
```

Comment:
It would be very interesting to learn about the advantages of this React-owned cache. What it can do that the user land's implementation can't implement? Or what it automates that a user land implementation would have to craft on their own?

Reply:
I don't know if this will be the final API for React Suspense caches. It's the one React DevTools has been using for a year or so now but it may change before public release.

React owning the underlying cache (map) allows it to manage invalidation (e.g. invalidate this bit of data and re-fetch it at a lower priority) without preventing it from re-rendering a high priority update in the meanwhile. If we used a global cache, it would be unable to do the high-priority update because the global cache would now be missing data.

In Replay's case, this doesn't matter much because we don't need to invalidate this cache for example. It grows over time but once added to it, data is never "wrong" or "outdated".